templates/company/contracts/CompanyTemplate.sol

```diff
@@ -48,24 +52,64 @@ contract CompanyTemplate is BaseTemplate {
+    )
+        public
+    {
+        _verifyCompanyParameters(_holders, _stakes, _votingSettings);
+        require(_payrollSettings.length == 4, ERROR_BAD_PAYROLL_SETTINGS);
+
+        (Kernel dao, ACL acl) = _createDAO();
+        (Finance finance, Voting voting) = _setupApps(dao, acl, _holders, _stakes, _votingSettings, _financePeriod, _useAgentAsVault);
+        _setupPayrollApp(dao, acl, _payrollSettings, finance, voting);
+        _transferRootPermissionsFromTemplate(dao, voting);
+        _registerID(_id, dao);
+    }
+
+    function _verifyCompanyParameters(address[] _holders, uint256[] _stakes, uint64[3] _votingSettings) internal {
         require(_holders.length > 0, ERROR_EMPTY_HOLDERS);
         require(_holders.length == _stakes.length, ERROR_BAD_HOLDERS_STAKES_LEN);
         require(_votingSettings.length == 3, ERROR_BAD_VOTE_SETTINGS);
+    }
+
+    function _setupApps(Kernel _dao, ACL _acl, address[] _holders, uint256[] _stakes, uint64[3] _votingSettings, uint64 _financePeriod, bool _useAgentAsVault) internal returns(Finance, Voting) {
         MiniMeToken token = _popTokenCache(msg.sender);
+        Vault agentOrVault = _useAgentAsVault ? _installDefaultAgentApp(_dao) : _installVaultApp(_dao);
+        Finance finance = _installFinanceApp(_dao, agentOrVault, _financePeriod == 0 ? DEFAULT_FINANCE_PERIOD : _financePeriod);
+        TokenManager tokenManager = _installTokenManagerApp(_dao, token, TOKEN_TRANSFERABLE, TOKEN_MAX_PER_ACCOUNT);
+        Voting voting = _installVotingApp(_dao, token, _votingSettings[0], _votingSettings[1], _votingSettings[2]);
 
-        // Create DAO and install apps
-        (Kernel dao, ACL acl) = _createDAO();
-        Vault agentOrVault = _useAgentAsVault ? _installDefaultAgentApp(dao) : _installVaultApp(dao);
-        Finance finance = _installFinanceApp(dao, agentOrVault, _financePeriod == 0 ? DEFAULT_FINANCE_PERIOD : _financePeriod);
-        TokenManager tokenManager = _installTokenManagerApp(dao, token, TOKEN_TRANSFERABLE, TOKEN_MAX_PER_ACCOUNT);
-        Voting voting = _installVotingApp(dao, token, _votingSettings[0], _votingSettings[1], _votingSettings[2]);
+        _mintTokens(_acl, tokenManager, _holders, _stakes);
+        _setupPermissions(_dao, _acl, agentOrVault, voting, finance, tokenManager, _useAgentAsVault);
 
-        // Mint tokens
-        _mintTokens(acl, tokenManager, _holders, _stakes);
+        return (finance, voting);
+    }
 
-        _setupPermissions(dao, acl, agentOrVault, voting, finance, tokenManager, _useAgentAsVault);
+    function _setupPayrollApp(Kernel _dao, ACL _acl, uint256[4] _payrollSettings, Finance _finance, Voting _voting) internal {
+        address denominationToken = _toAddress(_payrollSettings[0]);
+        IFeed priceFeed = IFeed(_toAddress(_payrollSettings[1]));
+        uint64 rateExpiryTime = _payrollSettings[2].toUint64();
+        address employeeManager = _toAddress(_payrollSettings[3]);
```

Comment:
One small thing is perhaps doing this coercion at the start of `newInstance()` instead, so the user doesn't waste gas creating an entire org before their input fails here.

Reply:
see my summary comment bellow